lib/blocks.php

```diff
@@ -296,3 +296,209 @@ function gutenberg_enqueue_editor_block_styles_assets() {
 	}
 	add_action( 'enqueue_block_editor_assets', 'gutenberg_enqueue_editor_block_styles_assets' );
 }
+
+/**
+ * Renders the classNames and styles for blocks
+ *
+ * @param string $block_content Output of the current block.
+ * @param array  $block Block Object.
+ * @return string New block output.
+ */
+function gutenberg_experimental_apply_classnames_and_styles( $block_content, $block ) {
+	if ( ! isset( $block['attrs'] ) ) {
+		return $block_content;
+	}
+
+	// Check what style features the block supports.
+	$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+	if ( ! $block_type ) {
+		return $block_content;
+	}
+
+	$supports = gutenberg_experimental_global_styles_get_supported_styles( $block_type->supports );
+
+	// Return early if nothing is supported.
+	if ( 0 === count( $supports ) ) {
+		return $block_content;
+	}
+
+	$colors     = gutenberg_experimental_build_css_colors( $block['attrs'], $supports );
+	$typography = gutenberg_experimental_build_css_typography( $block['attrs'], $supports );
+
+	$extra_classes = array_merge(
+		$colors['css_classes'],
+		$typography['css_classes'],
+		isset( $block['attrs']['className'] ) ? array( $block['attrs']['className'] ) : array(),
+		isset( $block['attrs']['align'] ) ? array( 'has-text-align-' . $block['attrs']['align'] ) : array()
```

Comment:
@youknowriad - I'm wondering if it makes sense for us to add the text alignment and className here?

If it does, should we skip the early return for no support flags and apply `align` and `className` for all blocks that have those attributes saved?  Or should we only apply these attrs if the block is using one of the experimental support flags?  (or not at all?)

<hr/>

Other than that, this PR should be ready for another look over once you get a chance. 😄

Reply:
My feeling is that this function knows too much about what each support flag (align, classnames, experimental ones)... does.

Ideally it's written in a generic way where we'd have a clear "interface" if want to add a new support flag. Doest that make sense?

Reply:
i seem something like

```
$attributes =  array();
$attributes = gutenberg_experimental_apply_colors_flag(  $attributes, $block['attrs'], $supports  );
$attributes = gutenberg_experimental_apply_lineheight_flag(  $attributes, $block['attrs'], $supports  );
$attributes = gutenberg_experimental_apply-align_flag(  $attributes, $block['attrs'], $supports  );
$attributes = gutenberg_experimental_apply_custom_classname_flag(  $attributes, $block['attrs'], $supports  );


if ( ! count( $attributes ) ) {
     return $post_content;
} 

// Do the magic to parse the html and add  the extra attributes (we could start by supporting class and style).
```

Reply:
I think that makes sense.

However, Im not currently seeing any of the similar experimental supports flags for the text alignment and class name.  Should we leave these out for now and expect them to be handled elsewhere for the time being (index.php I'm guessing for these blocks)?

Reply:
We have the `customClassName` for the custom classname, `classname`  for the generated classname (which can also be handled similarly) and `align` for text alligment. I'm fine leaving them for dedicated PRs though and just get the API right in this initial PR.

Reply:
> I'm fine leaving them for dedicated PRs though and just get the API right in this initial PR.

Sounds good.

> We have the customClassName for the custom classname, classname for the generated classname (which can also be handled similarly) and align for text alignment.

I saw these, but the reasons I am confused is that they don't have support keys defined for them like the others (font-size, line-height, colors, etc.) do.  Similarly, they aren't checked for in the `gutenberg_experimental_global_styles_get_supported_styles` function:
https://github.com/WordPress/gutenberg/blob/b98418d1c0c64fa4613c782afce2009fdb0779d5/lib/global-styles.php#L272-L278

So I was a bit confused about how they are supposed to be handled compared to the ones we have here that require the support flags, and if they are intended to be added here or elsewhere.  

Are we to assume these ones without support keys are supported by default if the corresponding attributes are saved and present?  If that is the case we can handle them that way here.

> align for text alignment

I was assuming this is for block alignment since it has settings like 'wide' and 'full' width.  Or does this combine both text alignment and block/element alignment?

Reply:
You're right "align" is for block alignment my bad but it's also a good candidate for server-side generation.

customClassName and className are more "historical" and not part of the global styles project but their behavior is very close. and yes, they are opt-out and I believe className can also receive a string to be used as class (need confirmation)

Reply:
> i seem something like

Updated to adopt this pattern instead.  Let me know if this isn't what you had in mind.